### PR TITLE
Feat/member form validation and redux wire-up

### PR DESCRIPTION
**Pull request by andregoncalves**

### CODE REVIEW

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 1]
```diff
@@ -1,64 +1,187 @@
...
+			attrs = fields.map(name => ({ name, value: this.state[name] }));
+		}
+		const errors = attrs.reduce(
+			(acc, curr) => {
+				const { name, value } = curr;
+				let isError;
+
+				if (['type', 'positions', 'shares', 'did'].includes(name)) {
+					const functionName =
+						'validateAttribute' + (name.charAt(0).toUpperCase() + name.substring(1));
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: this smells, better to do update `isValidAttribute`

if/switch/validators object by name will be good

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 2]
```diff
@@ -1,64 +1,187 @@
...
+	}
+
+	validateAttributeType = type => {
+		return ['individual', 'corporate'].includes(type);
 	};
 
-	closeAction = () => {
-		this.props.dispatch(push('/main/create-corporate-profile'));
+	validateAttributePositions = selectedPositions => {
+		const acceptablePositions = this.props.positions.map(p => p.position);
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: Instead of validating position, selected positions should be filtered by possible positions on change and before submit

File: src/common/identity/selectors.js [hunk 3]
```diff
@@ -387,6 +475,16 @@ export const selectCorporateLegalEntityTypes = createSelector(
 	idType => (idType ? idType.content.enum : [])
 );
 
+// Countries
+
+export const selectCountries = createSelector(
+	state => selectIdAttributeTypeByUrl(state, { attributeTypeUrl: COUNTRY_ATTRIBUTE }),
+	idType => {
+		const { enum: codes, enumNames: names } = idType.content.properties.country;
+		return codes.map((country, index) => ({ country, denonym: names[index] }));
```
Comment (sk91) [inline src/common/identity/selectors.js]: denonym is incorrect, should be "name"

denonym:
"The word used to identify citizens of the country"

For example: 
Country: "Germany"
Denonym: "German"

File: src/renderer/corporate/corporate-container.jsx [hunk 4]
```diff
@@ -23,7 +23,10 @@ class CorporateContainerComponent extends PureComponent {
 					path={`${match.path}/setup-corporate-profile/:identityId`}
 					component={CorporateWizardContainer}
 				/>
-				<Route path={`${match.path}/add-member`} component={CorporateAddMemberContainer} />
+				<Route
+					path={`${match.path}/add-member/:parentId?`}
+					component={CorporateAddMemberContainer}
```
Comment (sk91) [inline src/renderer/corporate/corporate-container.jsx]: better to use "render" instead of "component".  Component will cause component remount on each state render

Which in turn can be caused by unrelated state change.

Untill we nail memoized selectors correctly this might be a performance issue

File: src/renderer/corporate/dashboard/dashboard-container.jsx [hunk 5]
```diff
@@ -89,12 +91,26 @@ class CorporateDashboardContainer extends PureComponent {
...
 				/>
 			</React.Fragment>
 		);
 	}
 }
 
+const mapStateToProps = (state, props) => {
+	const profile = identitySelectors.selectCorporateProfile(state);
+	return {
+		identity: identitySelectors.selectIdentity(state),
```
Comment (sk91) [inline src/renderer/corporate/dashboard/dashboard-container.jsx]: profile should already include identity

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 6]
```diff
@@ -48,17 +77,26 @@ class CorporateAddMemberContainer extends PureComponent {
...
+			creationDate: 'Please enter company incorporation date',
+			taxId: 'Tax id provided is invalid',
+			type: 'Invalid member type',
+			positions: 'Invalid position',
+			country: 'Please select Residency',
+			nationality: 'Please select Nationality',
+			firstName: 'Please enter your first Name',
+			lastName: 'Please enter your last Name',
+			phoneNumber: 'Invalid phone number',
+			equity: 'Shares must be between 0 and 100',
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: equity could be of different types, not necessary shares. 
`CorporateStructureSchema.getEquityForCompanyType` can be used for this

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 7]
```diff
@@ -128,22 +216,35 @@ class CorporateAddMemberContainer extends PureComponent {
...
+
 	return {
-		parentIdentity,
-		basicAttributeTypes: identitySelectors.selectBasicCorporateAttributeTypes(state),
-		basicIdentity: identitySelectors.selectCorporateProfile(state, {
-			identityId: props.match.params.identityId
-		}),
+		parentId,
+		parentProfile,
+		identity,
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: shouldn't identity always point to `parentProfile.identity`
If so, `selectIdentity` should not be called, instead `selectCorporateProfile` should be called without props (orr with `{ identityId: null }`. It will fetch the current identity inside it's `identity` field

File: src/renderer/corporate/member/member-individual-form.jsx [hunk 8]
```diff
@@ -140,53 +140,53 @@ const CorporateMemberIndividualFormComponent = withStyles(styles)(props => {
...
 					>
 						<MenuItem value="">
 							<em>Choose...</em>
 						</MenuItem>
-						{jurisdictions.map(item => (
-							<MenuItem key={item} value={item}>
-								{item}
+						{countries.map(item => (
+							<MenuItem key={item.country} value={item}>
+								{item.denonym}
```
Comment (sk91) [inline src/renderer/corporate/member/member-individual-form.jsx]: denonym -> countryName

File: src/renderer/corporate/member/member-shares-form.jsx [hunk 9]
```diff
@@ -102,27 +101,27 @@ const CorporateMemberSharesFormComponent = withStyles(styles)(props => {
 			<div className={classes.inputWrap}>
 				{showParentCompany && (
 					<div className={`${classes.inputBox} ${classes.flexColumn}`}>
+						{console.log(companies)}
```
Comment (sk91) [inline src/renderer/corporate/member/member-shares-form.jsx]: stray log

File: src/common/identity/constants.js [hunk 10]
```diff
@@ -41,6 +42,24 @@ export const BASIC_CORPORATE_ATTRIBUTES = {
...
+export const CORPORATE_MEMBER_CORPORATE_ATTRIBUTES = {
+	[ENTITY_NAME_ATTRIBUTE]: 1,
+	[ENTITY_TYPE_ATTRIBUTE]: 1,
+	[JURISDICTION_ATTRIBUTE]: 1,
+	[EMAIL_ATTRIBUTE]: 1,
+	[CREATION_DATE_ATTRIBUTE]: 1,
+	[TAX_ID_ATTRIBUTE]: 1
+};
+
 export const CORPORATE_STRUCTURE =
```
Comment (sk91) [inline src/common/identity/constants.js]: lets make it `CORPORATE_STRUCTURE_ATTRIBUTE` so all will have same convention

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 11]
```diff
@@ -112,12 +199,13 @@ class CorporateAddMemberContainer extends PureComponent {
 	}
 
 	render() {
-		const membersForm = _.pick(this.state, 'errors');
+		const membersForm = _.pick(this.state, 'errors', this.selectFields(this.state.type));
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: does it work? as far as I know `_.pick` requires an array as second argument

File: src/renderer/corporate/member/corporate-add-member-container.jsx [hunk 12]
```diff
@@ -128,22 +216,35 @@ class CorporateAddMemberContainer extends PureComponent {
...
 		entityTypes: identitySelectors.selectCorporateLegalEntityTypes(state),
-		positions: identitySelectors.selectPositionsForCompanyType(state, {
-			companyType: parentIdentity.profile.entityType
-		})
+		availablePositions: identitySelectors.selectPositionsForCompanyType(state, {
+			companyType: parentProfile.entityType
+		}),
+		companies: [
+			parentProfile,
+			...parentProfile.members.filter(m => m.identity.type === 'corporate')
```
Comment (sk91) [inline src/renderer/corporate/member/corporate-add-member-container.jsx]: probably should be in the profile selector

File: src/renderer/corporate/corporate-container.jsx (hunk 4 above)
Comment (andregoncalves) [reply src/renderer/corporate/corporate-container.jsx]: Fixed on https://github.com/SelfKeyFoundation/Identity-Wallet/pull/1832/commits/61ddf2356740d6d2822b51c131d08a3863272fed

File: src/renderer/corporate/dashboard/dashboard-container.jsx (hunk 5 above)
Comment (andregoncalves) [reply src/renderer/corporate/dashboard/dashboard-container.jsx]: Good point.
Fixed on https://github.com/SelfKeyFoundation/Identity-Wallet/pull/1832/commits/82bcfef840644f20b6ebd7eb0cf8f7fe2e5f968d

File: src/renderer/corporate/member/corporate-add-member-container.jsx (hunk 7 above)
Comment (andregoncalves) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: I want to make the component flexible enough to be opened by different identity.
But it doesn't make sense, removed on https://github.com/SelfKeyFoundation/Identity-Wallet/pull/1832/commits/ab4f1f8df3920fec8b77f8adfb05399c32f2a40f

File: src/renderer/corporate/member/corporate-add-member-container.jsx (hunk 11 above)
Comment (andregoncalves) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: Screenshot 2019-11-19 at 16 35 55
Yikes! It works, that's why I didn't catch it. 
But it's undocumented, so I fixed it on https://github.com/SelfKeyFoundation/Identity-Wallet/pull/1832/commits/9a6b064081cc76426423b17776b0bb8e4499c16e

File: src/common/identity/selectors.js (hunk 3 above)
Comment (andregoncalves) [reply src/common/identity/selectors.js]: Fixed on f5f6884

File: src/renderer/corporate/member/corporate-add-member-container.jsx (hunk 6 above)
Comment (andregoncalves) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: Moving this fix to a different issue: #1865

File: src/renderer/corporate/member/corporate-add-member-container.jsx (hunk 12 above)
Comment (andregoncalves) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: What do you mean?
Have a new corporate profile attribute called companies, instead of aggregating them here?

Comment (sk91) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: yes, that what I meant, that way when we have memoization we will not need to filter.
But nvm for now if you do it only here

Comment (andregoncalves) [reply src/renderer/corporate/member/corporate-add-member-container.jsx]: Agree, we should strive to use memoization everywhere.
Improvement on 49e4583
